pipeline/frontend/yaml/compiler/compiler.go

```diff
@@ -328,3 +355,74 @@ func (c *Compiler) setupCacheRebuild(conf *yaml_types.Workflow, ir *backend_type
 
 	return nil
 }
+
+func dfsVisit(steps map[string]*stepWithDependsOn, name string, visited map[string]struct{}, path []string) error {
+	if _, ok := visited[name]; ok {
+		return fmt.Errorf("cycle detected: %v", path)
+	}
+
+	visited[name] = struct{}{}
+	path = append(path, name)
+
+	for _, dep := range steps[name].dependsOn {
+		if err := dfsVisit(steps, dep, visited, path); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func convertDAGToStages(steps map[string]*stepWithDependsOn) ([]*backend_types.Stage, error) {
```

Comment:
well I like to get rid of stages in the backend and only use a DAG independent of if we have it sequential or via group or via dag.

for now this should do but it wont do a full parallelism :/

Reply:
Considered and refactored it to that in the beginning as well, but I would actually stick to the "stages" representation as its pretty nice to check the dependencies this way.

Reply:
but it it will make a dag into "group"'s again and we loose the main feature of the dag so then we could just stick with groups

Reply:
Aniway that should not bloc p.k fronted changes as p. backend can be changed independent

Reply:
Sure, we can adjust that later on. Anything else missing for an approval?

Reply:
let me review it :)